core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java

```diff
@@ -0,0 +1,210 @@
+            } else {
+                return new SplitPartConstIndexFunction(strFunc, delimiterFunc, index);
+            }
+        }
+        return new SplitPartFunction(strFunc, delimiterFunc, indexFunc);
+    }
+
+    private static class SplitPartFunction extends AbstractSplitPartFunction implements TernaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final Function indexFunc;
+
+        public SplitPartFunction(Function strFunc, Function delimiterFunc, Function indexFunc) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.indexFunc = indexFunc;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return indexFunc.getInt(rec);
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getCenter() {
+            return delimiterFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return indexFunc;
+        }
+    }
+
+    private static class SplitPartConstIndexFunction extends AbstractSplitPartFunction implements BinaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final int index;
+
+        public SplitPartConstIndexFunction(Function strFunc, Function delimiterFunc, int index) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.index = index;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return index;
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return delimiterFunc;
+        }
+    }
+
+    private static abstract class AbstractSplitPartFunction extends StrFunction {
+        private final StringSink sink = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final Function strFunc;
+        private final Function delimiterFunc;
+
+        public AbstractSplitPartFunction(Function strFunc, Function delimiterFunc) {
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+        }
+
+        abstract int getIndex(Record rec);
+
+        @Override
+        public CharSequence getStr(Record rec) {
+            return getStr0(rec, sink);
+        }
+
+        @Override
+        public CharSequence getStrB(Record rec) {
+            return getStr0(rec, sinkB);
+        }
+
+        @Override
+        public void getStr(Record rec, CharSink sink) {
+            getStr0(rec, sink);
+        }
+
+        @Nullable
+        private <S extends CharSink> S getStr0(Record rec, S sink) {
```

Comment:
I see this pattern a lot in `StrFunction` implementations:
```java
@Override
public CharSequence getStr(Record rec) {
    return getStr0(rec, sink);}

@Override
public CharSequence getStrB(Record rec) {
     return getStr0(rec, sinkB);
}

@Override
public void getStr(Record rec, CharSink sink) {
    // ...
    sink.put(...);
}

@Nullable
private StringSink getStr0(Record rec, StringSink sink) {
    // ...
    sink.clear()
    sink.put(...);
}
```

This leads to a lot of code duplication between the `getStr(Record, CharSink)` and `getStr0(Record, StringSink)` functions, the only difference being that the `getStr0(Record, StringSink)` function calls `StringSink#clear()`.

To reduce code duplication, I made `getStr0(Record, CharSink)` generic. `getStr0(Record, CharSink)` also checks if the `CharSink` is an instance of `Mutable`, and if so, calls `Mutable.clear()` before putting the new value (`StringSink` implements `Mutable`).
```java
@Override
public CharSequence getStr(Record rec) {
    return getStr0(rec, sink);}

@Override
public CharSequence getStrB(Record rec) {
     return getStr0(rec, sinkB);
}

@Override
public void getStr(Record rec, CharSink sink) {
    getStr0(rec, sink);
}

@Nullable
private <S extends CharSink> S getStr0(Record rec, S sink) {
    // ...
    if (sink instanceof Mutable) {
        sink.clear();
    }
    sink.put(...);
}
```